pulsar-jms-tracing/src/main/java/com/datastax/oss/pulsar/jms/tracing/BrokerTracing.java

```diff
@@ -0,0 +1,484 @@
+
+import static com.datastax.oss.pulsar.jms.tracing.TracingUtils.TraceLevel;
+import static com.datastax.oss.pulsar.jms.tracing.TracingUtils.getConnectionDetails;
+import static com.datastax.oss.pulsar.jms.tracing.TracingUtils.getConsumerDetails;
+import static com.datastax.oss.pulsar.jms.tracing.TracingUtils.getEntryDetails;
+import static com.datastax.oss.pulsar.jms.tracing.TracingUtils.getMessageMetadataDetails;
+import static com.datastax.oss.pulsar.jms.tracing.TracingUtils.getProducerDetails;
+import static com.datastax.oss.pulsar.jms.tracing.TracingUtils.getPublishContextDetails;
+import static com.datastax.oss.pulsar.jms.tracing.TracingUtils.getSubscriptionDetails;
+import static com.datastax.oss.pulsar.jms.tracing.TracingUtils.trace;
+import static com.datastax.oss.pulsar.jms.tracing.TracingUtils.traceByteBuf;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import io.netty.buffer.ByteBuf;
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.bookkeeper.mledger.Entry;
+import org.apache.pulsar.broker.PulsarService;
+import org.apache.pulsar.broker.intercept.BrokerInterceptor;
+import org.apache.pulsar.broker.service.Consumer;
+import org.apache.pulsar.broker.service.Producer;
+import org.apache.pulsar.broker.service.ServerCnx;
+import org.apache.pulsar.broker.service.Subscription;
+import org.apache.pulsar.broker.service.Topic;
+import org.apache.pulsar.common.api.proto.BaseCommand;
+import org.apache.pulsar.common.api.proto.CommandAck;
+import org.apache.pulsar.common.api.proto.MessageMetadata;
+import org.apache.pulsar.common.intercept.InterceptException;
+
+@Slf4j
+public class BrokerTracing implements BrokerInterceptor {
+
+  public enum EventReasons {
+    ADMINISTRATIVE,
+    MESSAGE,
+    TRANSACTION,
+    SERVLET,
+  }
+
+  private static final TraceLevel defaultTraceLevel = TraceLevel.BASIC;
+
+  private static final LoadingCache<PulsarService, Set<EventReasons>> jmsTracingEventList =
+      CacheBuilder.newBuilder()
+          .maximumSize(10)
+          .expireAfterWrite(1, TimeUnit.MINUTES)
+          .build(
+              new CacheLoader<PulsarService, Set<EventReasons>>() {
+                public Set<EventReasons> load(PulsarService pulsar) {
+                  // dynamic config can get reloaded without service restart
+                  String events =
+                      pulsar
+                          .getConfiguration()
+                          .getProperties()
+                          .getProperty("jmsTracingEventList", "");
```

Comment:
this is not supposed to change during the lifetime of a broker, there is partial support for "dynamic properties" in Pulsar and IIRC you have to declare them somewhere.

is there any feature added recently for reloading the service configuration ?

I think that it is fine to keep this simple and require a broker restart